File: woggle.py
#! /usr/bin/env python3
# vim: ts=8:sts=8:sw=8:noexpandtab

import random
#from datetime import datetime
import unicodedata

# From http://www.bananagrammer.com/2013/10/the-boggle-cube-redesign-and-its-effect.html
# "q" == "QU"
dice = {
	"Classic" : [
		"AACIOT", "ABILTY", "ABJMOq", "ACDEMP",
		"ACELRS", "ADENVZ", "AHMORS", "BIFORX",
		"DENOSW", "DKNOTU", "EEFHIY", "EGKLUY",
		"EGINTV", "EHINPS", "ELPSTU", "GILRUW",
	],
	"New": [
		"AAEEGN", "ABBJOO", "ACHOPS", "AFFKPS",
		"AOOTTW", "CIMOTU", "DEILRX", "DELRVY",
		"DISTTY", "EEGHNW", "EEINSU", "EHRTVW",
		"EIOSST", "ELRTTY", "HIMNUq", "HLNNRZ",
	],
	"International": [
		"ETUKNO", "EVGTIN", "DECAMP", "IELRUW",
		"EHIFSE", "RECALS", "ENTDOS", "OFXRIA",
		"NAVEDZ", "EIOATA", "GLENYU", "BMAQJO",
		"TLIBRA", "SPULTE", "AIMSOR", "ENHRIS",
	],
}

def buildref(file):
	#t = datetime.now()
	ref = dict()
	numwords = 0
	with open(file, "r") as words_f:
		for line in words_f:
			word = unicodedata.normalize("NFD", line.strip().upper()) \
				.encode("ascii", "ignore").decode("utf-8")
			length = len(word)
			if length <= 3:
				continue
			_ref = ref
			for c in word:
				if c not in _ref:
					_ref[c] = dict()
				_ref = _ref[c]
			if True not in _ref:
				_ref[True] = "%"
				numwords += 1
			if numwords % 10000 == 0:
				print(numwords, end="\r")
	#ms = (datetime.now() - t).total_seconds() * 1000
	#print(f"Built dict containing {numwords} words in {ms:g}ms")
	return ref, numwords

def inref(ref, word):
	_ref = ref
	for c in word:
		if c not in _ref:
			return None
		_ref = _ref[c]
	if True in _ref:
		return True
	return False

#print(f"inref('aggression') = {inref('aggression'.upper())}")

def randgrid(diceset):
	grid = ["#" for i in range(16)]
	thrown = list()
	for die in dice[diceset]:
		thrown.append(random.choice(die))
	for i in range(4):
		for j in range(4):
			t = random.randint(0, len(thrown)-1)
			grid[i*4+j] = thrown[t]
			del thrown[t]
	assert len(thrown) == 0
	return grid

def printgrid(grid, path=None, good=None, bad=None):
	for i in range(4):
		for j in range(4):
			c = grid[i*4+j].replace("q", "QU")
			fmt = ""
			if good is not None \
			and (i,j) in good:
				if path is not None \
				and (i,j) in path:
					fmt = "1;42;30"
				else:
					fmt = "42;30"
			elif bad is not None \
			and (i,j) in bad:
				if path is not None \
				and (i,j) in path:
					fmt = "1;41;30"
				else:
					fmt = "41;30"
			print(f"\033[{fmt}m{c:3}\033[m", end="")
		print()
	print()

word = ""
def _findwords(ref, grid, done, found, pos):
	i, j = pos
	global word
	assert not done[i*4+j]
	done[i*4+j] = True
	word += grid[i*4+j]
	_word = word.replace("q", "QU")
	ret = inref(ref, _word)
	if len(_word) > 3 \
	and ret == True \
	and _word not in found:
		found.append(_word)
	if ret is not None:
		for ii in range(max(0,i-1),min(4,i+2)):
			for jj in range(max(0,j-1),min(4,j+2)):
				if not done[ii*4+jj]:
					_findwords(ref, grid, done, found, (ii,jj))
	word = word[:-1]
	done[i*4+j] = False

def findwords(ref, grid):
	#t = datetime.now()
	found = []
	done = [False] * 16
	global word
	word = ""
	for i in range(4):
		for j in range(4):
			_findwords(ref, grid, done, found, (i,j))
	#ms = (datetime.now() - t).total_seconds() * 1000
	#print(f"Found {len(found)} words in {ms:g}ms")
	return found

def updatefound(grid, found, good, bad):
	newfound = []
	for word in found:
		path = set(findpath(grid, word))
		if good.issubset(path) \
		and not path.intersection(bad):
			newfound.append(word)
	return newfound

def printwords(found):
	maxlength = 0
	for word in found:
		maxlength = max(len(word), maxlength)
	curlen = 0
	for word in found:
		print(f"{word:<{maxlength}} ", end="")
		curlen += maxlength + 1
		if curlen > 80:
			curlen = 0
			print()
	if curlen > 0:
		print()

def _findpath(grid, done, word, pos):
	i, j = pos
	assert not done[i*4+j]
	assert len(word) > 0
	qu = (grid[i*4+j] == "q" and word[:2] == "QU")
	if grid[i*4+j] != word[0] and not qu:
		return None
	if len(word) == 1:
		return [(i,j)]
	done[i*4+j] = True
	for ii in range(max(0,i-1),min(4,i+2)):
		for jj in range(max(0,j-1),min(4,j+2)):
			if done[ii*4+jj]:
				continue
			if qu:
				ret = _findpath(grid, done, word[2:], (ii,jj))
			else:
				ret = _findpath(grid, done, word[1:], (ii,jj))
			if ret is not None:
				done[i*4+j] = False
				return [(ii,jj)] + ret
	done[i*4+j] = False
	return None

def findpath(grid, word):
	for i in range(4):
		for j in range(4):
			done = [False] * 16
			ret = _findpath(grid, done, word, (i,j))
			if ret is not None:
				return [(i,j)] + ret
	return None

print("EN/FR?")
while True:
	lang = input("> ").upper()
	if lang == "EN":
		ref, numwords = buildref("scrabble-dictionary/dictionary.csv")
		diceset = "Classic"
	elif lang == "FR":
		ref, numwords = buildref("liste.de.mots.francais.frgut.txt")
		diceset = "International"
	else:
		print("invalid language")
		continue
	break
print(f"The dictionary contains {numwords} words (≥ 4 letters)" if lang == "EN" \
 else f"Le dictionnaire contient {numwords} mots (≥ 4 lettres)")
while True:
	while True:
		grid = randgrid(diceset)
		found = findwords(ref, grid)
		if len(found) > 30:
			break
	print(f"There are {len(found)} words in the grid" if lang == "EN" \
	 else f"Il y a {len(found)} mots dans la grille")
	printgrid(grid)
	target = random.choice(found)
	targetpath = findpath(grid, target)
	good = set()
	bad = set()
	n = 1
	while n <= 5:
		word = input(f"{n}/5 > ").upper()
		word = unicodedata.normalize("NFD", word) \
			.encode("ascii", "ignore").decode("utf-8")
		if word == "":
			printgrid(grid, path, good, bad)
			continue
		if word == "SPOIL ME":
			printwords(found)
			continue
		if word == "SPOIL ME THE TARGET":
			print(target)
			continue
		path = findpath(grid, word)
		if path is None:
			print(f"{word} is not in the grid" if lang == "EN" \
			 else f"{word} n'est pas dans la grille", end=" ")
			continue
		if inref(ref, word) != True:
			print(f"{word} is not in the dictionary" if lang == "EN" \
			 else f"{word} n'est pas dans le dictionnaire", end=" ")
			continue
		for pos in path:
			if pos in targetpath:
				good.add(pos)
			else:
				bad.add(pos)
		printgrid(grid, path, good, bad)
		if word == target:
			print(f"you win in {n} round{'s' if n > 1 else ''}!!!" if lang == "EN" \
			 else f"tu gagnes en {n} tour{'s' if n > 1 else ''} !!!")
			break
		n += 1
		if word in found:
			found.remove(word)
		found = updatefound(grid, found, good, bad)
		print(f"{len(found)} words possible left" if lang == "EN" \
		 else f"Il reste {len(found)} mots possibles")
	else:
		print(f"you loose... the word was {target}" if lang == "EN" \
		 else f"tu as perdu ... le mot était {target}")
	print()
	print("play again? (Y/n)" if lang == "EN" \
	 else "rejouer ? (O/n)")
	cmd = input("> ")
	if cmd in ["n", "N"]:
		break

File: woggle.js
import {dic} from "./fr.dic.js";

function buildref() {
	let ref = {};
	let numwords = 0;
	let pos = [];
	let _ref = ref;
	for (let c of dic) {
		if (c == ".") {
			_ref[true] = true;
			if (++numwords % 10000 == 0)
				console.log(numwords);
		}
		else if (c == ";")
			_ref = pos.pop();
		else {
			_ref[c] = {};
			pos.push(_ref);
			_ref = _ref[c];
		}
	}
	//assert(ref == _ref);
	return [ref, numwords];
}

function inref(ref, word) {
	let _ref = ref;
	for (let c of word) {
		if (!_ref[c])
			return null;
		_ref = _ref[c];
	}
	if (_ref[true])
		return true;
	return false;
}

// From http://www.bananagrammer.com/2013/10/the-boggle-cube-redesign-and-its-effect.html
// "q" == "QU"
const dice = {
	"Classic": [
		"AACIOT", "ABILTY", "ABJMOq", "ACDEMP",
		"ACELRS", "ADENVZ", "AHMORS", "BIFORX",
		"DENOSW", "DKNOTU", "EEFHIY", "EGKLUY",
		"EGINTV", "EHINPS", "ELPSTU", "GILRUW",
	],
	"New": [
		"AAEEGN", "ABBJOO", "ACHOPS", "AFFKPS",
		"AOOTTW", "CIMOTU", "DEILRX", "DELRVY",
		"DISTTY", "EEGHNW", "EEINSU", "EHRTVW",
		"EIOSST", "ELRTTY", "HIMNUq", "HLNNRZ",
	],
	"International": [
		"ETUKNO", "EVGTIN", "DECAMP", "IELRUW",
		"EHIFSE", "RECALS", "ENTDOS", "OFXRIA",
		"NAVEDZ", "EIOATA", "GLENYU", "BMAQJO",
		"TLIBRA", "SPULTE", "AIMSOR", "ENHRIS",
	],
};

function randgrid(diceset) {
	// Initialize grid
	let grid = [];
	for (let g=0; g<16; ++g)
		grid.push("#");
	// Throw each dice
	let thrown = []
	for (let d=0; d<dice[diceset].length; ++d) {
		let n = Math.floor(Math.random() * 6);
		thrown.push(dice[diceset][d].charAt(n));
	}
	// Dice permutation
	for (let g=0; g<16; ++g) {
		let n = Math.floor(Math.random() * thrown.length);
		grid[g] = thrown[n];
		thrown.splice(n, 1);
	}
	//assert len(thrown) == 0
	return grid;
}

function _findwords(ref, grid, done, found, word, i, j) {
	//assert not done[i*4+j]
	done[i*4+j] = true;
	word += grid[i*4+j];
	let _word = word.replaceAll("q", "QU");
	let ret = inref(ref, _word);
	if (ret === true && !found.includes(_word))
		found.push(_word);
	if (ret !== null) {
		for (let ii=Math.max(0,i-1); ii<Math.min(4,i+2); ++ii) {
		for (let jj=Math.max(0,j-1); jj<Math.min(4,j+2); ++jj) {
			if (!done[ii*4+jj])
				_findwords(ref, grid, done, found, word, ii, jj);
		}
		}
	}
	done[i*4+j] = false;
}

function findwords(ref, grid) {
	let found = [];
	let done = [];
	for (let i=0; i<16; ++i)
		done.push(false);
	for (let i=0; i<4; ++i) {
	for (let j=0; j<4; ++j)
		_findwords(ref, grid, done, found, "", i, j);
	}
	return found;
}

function updatecells(grid) {
	for (let g=0; g<16; ++g) {
		let cell = document.querySelector(`#cell_${g}`);
		let c = grid[g];
		if (c == "q")
			cell.textContent = "Q";
		else
			cell.textContent = c;
	}
}

let ref, grid, found;
let valid = [];
let curpath = [];
let word = "";
let counter = 0;

function allvalid() {
	for (let i=0; i<4; ++i)
	for (let j=0; j<4; ++j)
		valid.push([i,j]);
}

function update() {
	if (curpath.length > 0) {
		let nextvalid = [];
		let [i,j] = curpath[curpath.length - 1];
		for (let ii=Math.max(0,i-1); ii<Math.min(4,i+2); ++ii) {
		for (let jj=Math.max(0,j-1); jj<Math.min(4,j+2); ++jj) {
			let used = false;
			for (let [ci,cj] of curpath) {
				if (ii==ci && jj==cj) {
					used = true;
					break;
				}
			}
			if (!used)
				nextvalid.push([ii,jj]);
		}
		}
		valid = nextvalid;
	} else
		allvalid();
	let disp = document.querySelector("#word");
	disp.textContent = word + "␣".repeat(Math.max(4-word.length, 1));
}

function addpos(i, j) {
	let g = i*4+j;
	console.log(`addpos(${grid[g]})`);
	word += grid[g].replace("q", "QU");
	if (word.length >= 4) {
		let enter = document.querySelector("#enter");
		enter.onclick = (event) => { play(); };
	}
	curpath.push([i,j]);
	update();
}

function removefrompos(i, j) {
	console.log(`removefrompos(${i},${j})`);
	for (let n = curpath.length - 1; n >= 0; --n) {
		let [ii,jj] = curpath[n];
		curpath.splice(n, 1);
		let c = grid[ii*4+jj].replace("q", "QU");
		let m = word.lastIndexOf(c);
		word = word.slice(0, m);
		if (ii==i && jj==j)
			break;
	}
	update();
}

function play() {
	if (inref(ref, word) === true) {
		incrementcounter();
	} else {
		word = "";
		curpath = [];
		update();
	}
} 

function incrementcounter() {
	++counter;
	let disp = document.querySelector("#counter");
	disp.textContent = `${counter}/5`;
}

window.onload = (event) => {
	const tbody = document.querySelector("table#grid > tbody");
	for (let i=0; i<4; ++i) {
		let row = document.createElement("tr");
		tbody.appendChild(row);
		for (let j=0; j<4; ++j) {
			let cell = document.createElement("td");
			cell.id = `cell_${i*4+j}`;
			row.appendChild(cell);
		}
	}
	let numwords;
	[ref, numwords] = buildref();
	console.log(`There are ${numwords} words in the dictionary`);
	while (true) {
		grid = randgrid("Classic");
		found = findwords(ref, grid);
		if (found.length > 30)
			break;
		console.log(`Only ${found.length} words in the grid, throwing the dice again`);
	}
	console.log(`There are ${found.length} words in the grid`);
	updatecells(grid);
	update();
	for (let g=0; g<16; ++g) {
		let cell = document.querySelector(`#cell_${g}`);
		cell.onclick = (event) => {
			console.log(`curpath=${curpath}`);
			for (let [i,j] of curpath) {
				if (i*4+j == g) {
					removefrompos(i,j);
					return;
				}
			}
			console.log(`valid=${valid}`);
			for (let [i,j] of valid) {
				if (i*4+j == g) {
					addpos(i, j);
					return;
				}
			}
			console.log("not in curpath nor valid");
			// show help?
		}
	}
	incrementcounter();
};
